pkg/sink/codec/bootstraper.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		return errors.Trace(err)
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case b.outCh <- event:
+	}
+	return nil
+}
+
+func (b *bootstrapWorker) generateEvents(
+	topic string,
+	tableInfo *model.TableInfo,
+) (*future, error) {
+	// Bootstrap messages of a table should be sent to all partitions.
+	msg, err := b.encoder.EncodeDDLEvent(model.NewBootstrapDDLEvent(tableInfo))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	key := model.TopicPartitionKey{
+		Topic:     topic,
+		Partition: defaultBootstrapPartitionIndex,
+	}
+
+	f := &future{
+		Key:  key,
+		done: make(chan struct{}),
+	}
+	f.Messages = append(f.Messages, msg)
+	close(f.done)
+	return f, nil
+}
+
+func (b *bootstrapWorker) gcInactiveTables() {
+	b.activeTables.Range(func(key, value interface{}) bool {
+		table := value.(*tableStatistic)
+		if table.isInactive(b.maxInactiveDuration) {
+			log.Info("A table is removed from the bootstrap worker",
+				zap.Int64("tableID", table.id),
+				zap.String("topic", table.topic),
+				zap.Stringer("changefeed", b.changefeedID))
+			b.activeTables.Delete(key)
+		}
+		return true
+	})
+}
+
+// tableStatistic is used to record the statistics of a table
+type tableStatistic struct {
+	// id is the table's ID, it will not change
+	id int64
+	// topic is the table's topic, it will not change
+	topic string
+	// counter is the number of row event sent since last bootstrap message sent
+	// It is used to check if the bootstrap message should be sent
+	counter atomic.Int32
+	// lastMsgReceivedTime is the last time the row event is received
+	// It is used to check if the table is inactive
+	lastMsgReceivedTime atomic.Value
+	// lastSendTime is the last time the bootstrap message is sent
+	// It is used to check if the bootstrap message should be sent
+	lastSendTime atomic.Value
+	// version is the table version
+	// It is used to check if the table schema is changed since last bootstrap message sent
+	version atomic.Uint64
+	// tableInfo is the tableInfo of the table
+	// It is used to generate bootstrap message
+	tableInfo atomic.Value
+}
+
+func newTableStatus(key model.TopicPartitionKey, row *model.RowChangedEvent) *tableStatistic {
+	res := &tableStatistic{
+		id:    row.Table.TableID,
+		topic: key.Topic,
+	}
+	res.counter.Add(1)
+	res.lastMsgReceivedTime.Store(time.Now())
+	res.lastSendTime.Store(time.Unix(0, 0))
+	res.version.Store(row.TableInfo.UpdateTS)
+	res.tableInfo.Store(row.TableInfo)
+	return res
+}
+
+func (t *tableStatistic) shouldSendBootstrapMsg(
+	sendBootstrapInterval time.Duration,
+	sendBootstrapMsgCountInterval int32,
+) bool {
+	lastSendTime := t.lastSendTime.Load().(time.Time)
+	return time.Since(lastSendTime) >= sendBootstrapInterval ||
+		t.counter.Load() >= sendBootstrapMsgCountInterval
+}
+
+func (t *tableStatistic) update(row *model.RowChangedEvent) {
+	t.counter.Add(1)
+	t.lastMsgReceivedTime.Store(time.Now())
+
+	if t.version.Load() != row.TableInfo.UpdateTS {
```

Comment:
send a  bootstrap message if the table version is changed.

Reply:
This is unneeded because we've already sent DDL event.